pdf/extractor/image.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * This file is subject to the terms and conditions defined in
+ * file 'LICENSE.md', which is part of this source code package.
+ */
+
+package extractor
+
+import (
+	"github.com/unidoc/unidoc/common"
+	"github.com/unidoc/unidoc/pdf/contentstream"
+	"github.com/unidoc/unidoc/pdf/core"
+	"github.com/unidoc/unidoc/pdf/model"
+)
+
+// ExtractPageImages returns the image contents of the page extractor, including data
+// and position, size information for each image.
+func (e *Extractor) ExtractPageImages() (*PageImages, error) {
+	ctx := &imageExtractContext{}
+
+	err := ctx.extractImagesInContentStream(e.contents, e.resources)
+	if err != nil {
+		return nil, err
+	}
+
+	images := &PageImages{
+		Images: ctx.extractedImages,
+	}
+	return images, nil
+}
+
+// PageImages represents extracted images on a PDF page with spatial information:
+// display position and size.
+type PageImages struct {
+	Images []ImageMark
+}
+
+// ImageMark represents an image drawn on a page and its position in device coordinates.
+// All coordinates are in device coordinates.
+type ImageMark struct {
+	Image *model.Image
+
+	// Dimensions of the image as displayed in the PDF.
+	Width  float64
+	Height float64
+
+	// Position of the image in PDF coordinates (lower left corner).
+	X float64
+	Y float64
+
+	// Angle if rotated.
+	Angle float64
+}
+
+// Provide context for image extraction content stream processing.
+type imageExtractContext struct {
+	extractedImages []ImageMark
+	inlineImages    int
+	xObjectImages   int
+	xObjectForms    int
+
+	// Cache to avoid processing same image many times.
+	cacheXObjectImages map[*core.PdfObjectStream]*cachedImage
+}
+
+type cachedImage struct {
+	image *model.Image
+	cs    model.PdfColorspace
+}
+
+func (ctx *imageExtractContext) extractImagesInContentStream(contents string, resources *model.PdfPageResources) error {
+	cstreamParser := contentstream.NewContentStreamParser(contents)
+	operations, err := cstreamParser.Parse()
+	if err != nil {
+		return err
+	}
+
+	if ctx.cacheXObjectImages == nil {
+		ctx.cacheXObjectImages = map[*core.PdfObjectStream]*cachedImage{}
+	}
+
+	processor := contentstream.NewContentStreamProcessor(*operations)
+	processor.AddHandler(contentstream.HandlerConditionEnumAllOperands, "",
+		func(op *contentstream.ContentStreamOperation, gs contentstream.GraphicsState, resources *model.PdfPageResources) error {
+			return ctx.processOperand(op, gs, resources)
+		})
+
+	err = processor.Process(resources)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Process individual content stream operands for image extraction.
+func (ctx *imageExtractContext) processOperand(op *contentstream.ContentStreamOperation, gs contentstream.GraphicsState, resources *model.PdfPageResources) error {
```

Comment:
This method seems too long to me. May be better to split it in separate methods. Something like:

```
func (ctx *imageExtractContext) processOperand(op *contentstream.ContentStreamOperation, gs contentstream.GraphicsState, resources *model.PdfPageResources) error {
    lenParams := len(op.Params)
    if op.Operand == "BI" && lenParams == 1 {
        iimg, ok := op.Params[0].(*contentstream.ContentStreamInlineImage)
        if !ok {
            return nil
        }

        return ctx.extractInlineImage(iimg, gs, resources)
    } else if op.Operand == "Do" && lenParams == 1 {
        name, ok := core.GetName(op.Params[0])
        if !ok {
            return errTypeCheck
        }

        _, xtype := resources.GetXObjectByName(*name)
        if xtype == model.XObjectTypeImage {
            return ctx.extractXObjectImage(name, gs, resources)
        } else if xtype == model.XObjectTypeForm {
            return ctx.extractFormImages(name, gs, resources)
        }
    }

    return nil
}
```